File: src/_687/Solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var longestUnivaluePath = function(root) {
    var ans = 0;

    function arrowLength(node) {
        var left, right, arrowLeft = 0, arrowRight = 0;

        if (node === null) return 0;
        left = arrowLength(node.left);
        right = arrowLength(node.right);
        if (node.left && node.left.val === node.val) arrowLeft = left + 1;
        if (node.right && node.right.val === node.val) arrowRight = right + 1;
        if (arrowLeft + arrowRight > ans) ans = arrowLeft + arrowRight;
        
        return Math.max(arrowLeft, arrowRight);
    }

    arrowLength(root);

    return ans;
};

File: src/_687/Solution.java
package _687;

class Solution {
    private int maxLen = 0;

    public int longestUnivaluePath(TreeNode root) {
        if (root == null) {
            return 0;
        }
        dfs(root, root.val + 1);
        return maxLen;
    }

    private int dfs(TreeNode node, int rootVal) {
        if (node == null) return 0;

        int left = dfs(node.left, node.val);
        int right = dfs(node.right, node.val);

        if (node.val == rootVal) {
            return Math.max(left, right) + 1;
        }

        maxLen = Math.max(maxLen, left + right);

        return 0;
    }

    private class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }
}